src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_StringConcat.cs

```diff
@@ -416,90 +343,96 @@ private BoundExpression RewriteStringConcatInExpressionLambda(SyntaxNode syntax,
+                    {
+                        return BoundCall.Synthesized(expr.Syntax, expr, toStringMethod);
                     }
                 }
             }
 
-            // Optimization not possible; just return the original expression.
-            return expr;
-        }
+            // If it's a value type (or unconstrained generic), and it's not constant and not readonly,
+            // then we need a copy. This is to mimic the old behaviour, where ToString was called on a box
+            // of the value type, and so any side-effects of ToString weren't made to the original.
+            if (!expr.Type.IsReferenceType && !expr.Type.IsReadOnly && expr.ConstantValue == null)
+            {
+                expr = new BoundPassByCopy(expr.Syntax, expr, expr.Type);
+            }
 
-        /// <summary>
-        /// Gets whether the type of an argument used in string concatenation can
-        /// be optimized by first calling ToString on it before passing the argument
-        /// to the String.Concat function.
-        /// </summary>
-        /// <param name="symbol">The type symbol of the argument.</param>
-        /// <returns>
-        /// true if ToString may be used; false if using ToString could lead to observable differences in behavior.
-        /// </returns>
-        private static bool ConcatExprCanBeOptimizedWithToString(TypeSymbol symbol)
-        {
-            // There are several constraints applied here in support of backwards compatibility:
-            // - This optimization potentially changes the order in which ToString is called
-            //   on the arguments.  That's a a compatibility issue if one argument's ToString
-            //   depends on state mutated by another, such as current culture.
-            // - For value types, this optimization causes ToString to be called on the original
-            //   value rather than on a boxed copy.  That means a mutating ToString implementation
-            //   could change the original rather than the copy.
-            // For these reasons, this optimization is currently restricted to primitives
-            // known to have a non-mutating ToString implementation that is independent
-            // of externally mutable state.  Common value types such as Int32 and Double
-            // do not meet this bar.
-
-            switch (symbol.SpecialType)
+            // No need for a conditional access if it's a value type - we know it's not null.
+            if (expr.Type.IsValueType)
             {
-                case SpecialType.System_Boolean:
-                case SpecialType.System_Char:
-                case SpecialType.System_IntPtr:
-                case SpecialType.System_UIntPtr:
-                    return true;
-                default:
-                    return false;
+                return BoundCall.Synthesized(expr.Syntax, expr, toString);
             }
+
+            int currentConditionalAccessID = ++_currentConditionalAccessID;
```

Comment:
I saw [one place in code where this was `++_currentConditionalAccessID`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter_ConditionalAccess.cs,73), and [another where it was `_currentConditionalAccessID++`](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp/Lowering/LocalRewriter/LocalRewriter_FixedStatement.cs,339). I'm assuming that the second is buggy (although I tried and failed to trigger it), and copied the pattern from the first.

Reply:
Anyone able to comment about why the increments vary in this way in different locations?

Reply:
@VSadov? You wrote the [original `_currentConditionalAccessId++`](https://github.com/dotnet/roslyn/commit/7cb61a28144b237b72f474a182cf674500f65a98).

Reply:
No, but can you please change it so that they are all preincrement?

---
In reply to: [278340949](https://github.com/dotnet/roslyn/pull/35006#discussion_r278340949) [](ancestors = 278340949)